File: src/tinymonitor/alerts/webhook.py
import requests
import json
import socket
from datetime import datetime
from tinymonitor.alerts.provider import AlertProvider, logger
from tinymonitor.utils.system_info import get_private_ip, get_public_ip, get_uptime, get_load_avg

class WebhookAlert(AlertProvider):
    def send(self, component, level, value, title, message):
        url = self.config.get("url")
        if not url:
            logger.error(f"[{self.name}] No url provided")
            return

        # Custom headers (e.g. Authorization)
        headers = self.config.get("headers", {})
        if "Content-Type" not in headers:
            headers["Content-Type"] = "application/json"

        timeout = self.config.get("timeout", 10)

        # Standardized JSON Payload
        payload = {
            "timestamp": datetime.now().isoformat(),
            "alert": {
                "level": level,
                "component": component,
                "value": str(value),
                "title": title,
                "message": message
            },
            "host": {
                "hostname": socket.gethostname(),
                "ip_private": get_private_ip(),
                "ip_public": get_public_ip(),
                "uptime": get_uptime(),
                "load_average": get_load_avg()
            }
        }

        try:
            response = requests.post(
                url, 
                json=payload, 
                headers=headers, 
                timeout=timeout
            )
            
            if response.ok: # Status 200-299
                logger.info(f"[{self.name}] Alert sent successfully")
            else:
                logger.error(f"[{self.name}] Failed to send alert: {response.status_code} {response.text}")
        except Exception as e:
            logger.error(f"[{self.name}] Error sending alert: {e}")

File: tinymonitor-go/internal/alerts/webhook.go
package alerts

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/Gu1llaum-3/tinymonitor/internal/config"
	"github.com/Gu1llaum-3/tinymonitor/internal/models"
	"github.com/Gu1llaum-3/tinymonitor/internal/utils"
)

// WebhookProvider sends alerts to a generic webhook
type WebhookProvider struct {
	BaseProvider
	url     string
	headers map[string]string
	timeout int
}

// NewWebhookProvider creates a new webhook provider
func NewWebhookProvider(cfg config.WebhookConfig) *WebhookProvider {
	return &WebhookProvider{
		BaseProvider: BaseProvider{
			ProviderName: "webhook",
			Enabled:      cfg.Enabled,
			Levels:       cfg.Levels,
			Rules:        cfg.Rules,
		},
		url:     cfg.URL,
		headers: cfg.Headers,
		timeout: cfg.Timeout,
	}
}

// Send sends an alert to the webhook
func (p *WebhookProvider) Send(alert models.Alert) error {
	if p.url == "" {
		return fmt.Errorf("no url provided")
	}

	payload := map[string]interface{}{
		"timestamp": time.Now().Format(time.RFC3339),
		"alert": map[string]interface{}{
			"level":     alert.Level,
			"component": alert.Component,
			"value":     alert.Value,
			"title":     alert.Title,
			"message":   alert.Message,
		},
		"host": map[string]interface{}{
			"hostname":     utils.GetHostname(),
			"ip_private":   utils.GetPrivateIP(),
			"ip_public":    utils.GetPublicIP(),
			"uptime":       utils.GetUptime(),
			"load_average": utils.GetLoadAvg(),
		},
	}

	body, err := json.Marshal(payload)
	if err != nil {
		return err
	}

	req, err := http.NewRequest("POST", p.url, bytes.NewReader(body))
	if err != nil {
		return err
	}

	// Set headers
	hasContentType := false
	for key, value := range p.headers {
		req.Header.Set(key, value)
		if key == "Content-Type" {
			hasContentType = true
		}
	}
	if !hasContentType {
		req.Header.Set("Content-Type", "application/json")
	}

	timeout := p.timeout
	if timeout <= 0 {
		timeout = 10
	}

	client := &http.Client{Timeout: time.Duration(timeout) * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("failed to send alert: status %d", resp.StatusCode)
	}

	LogInfo(p.ProviderName, "Alert sent successfully")
	return nil
}
